src/libcore/num/mod.rs

```diff
@@ -112,6 +112,112 @@ nonzero_integers! {
+#[unstable(feature = "nonzero_parse", issue = "0")]
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum NonZeroIntErrorKind {
+    /// Value being parsed is empty.
+    ///
+    /// Among other causes, this variant will be constructed when parsing an empty string.
+    Empty,
+    /// Contains an invalid digit.
+    ///
+    /// Among other causes, this variant will be constructed when parsing a string that
+    /// contains a letter.
+    InvalidDigit,
+    /// Integer is too large to store in target integer type.
+    Overflow,
+    /// Integer is too small to store in target integer type.
+    Underflow,
+    /// Integer contains the value `0` which is forbidden for a non-zero integer
+    Zero,
+}
+
+#[unstable(feature = "nonzero_parse", issue = "0")]
+impl From<ParseIntError> for ParseNonZeroIntError {
+    fn from(p: ParseIntError) -> Self {
+        use self::IntErrorKind as IK;
+        use self::NonZeroIntErrorKind as NK;
+        ParseNonZeroIntError {
+            kind: match p.kind {
+                IK::Empty => NK::Empty,
+                IK::InvalidDigit => NK::InvalidDigit,
+                IK::Overflow => NK::Overflow,
+                IK::Underflow => NK::Underflow,
+            },
+        }
+    }
+}
+
+impl ParseNonZeroIntError {
+    /// Outputs the detailed cause of parsing an integer failing.
+    #[unstable(feature = "int_error_matching",
+               reason = "it can be useful to match errors when making error messages \
+                         for integer parsing",
+               issue = "22639")]
+    pub fn kind(&self) -> &NonZeroIntErrorKind {
+        &self.kind
+    }
+
+    #[unstable(feature = "int_error_internals",
+               reason = "available through Error trait and this method should \
+                         not be exposed publicly",
+               issue = "0")]
+    #[doc(hidden)]
+    pub fn __description(&self) -> &str {
+        match self.kind {
+            NonZeroIntErrorKind::Empty => "cannot parse integer from empty string",
+            NonZeroIntErrorKind::InvalidDigit => "invalid digit found in string",
+            NonZeroIntErrorKind::Overflow => "number too large to fit in target type",
+            NonZeroIntErrorKind::Underflow => "number too small to fit in target type",
+            NonZeroIntErrorKind::Zero => "number is 0",
+        }
+    }
+
+}
+
+#[unstable(feature = "nonzero_parse", issue = "0")]
+impl fmt::Display for ParseNonZeroIntError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.__description().fmt(f)
+    }
+}
+
+
+macro_rules! from_str_radix_nzint_impl {
+    ($($t:ty)*) => {$(
+        #[unstable(feature = "nonzero_parse", issue = "0")]
+        impl FromStr for $t {
+            type Err = ParseNonZeroIntError;
+            fn from_str(src: &str) -> Result<Self, Self::Err> {
+                Self::new(from_str_radix(src, 10)?)
```

Comment:
These `FromStr` impls are already for convenience rather than strictly necessary. I think parsing non-10 bases for non-zero types gets niche enough that it’s probably not worth dedicated API surface. You can always use e.g. `u32::from_str_radix` then `NonZeroU32::new`.